src/boot/efi/devicetree.c

```diff
@@ -0,0 +1,158 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <efi.h>
+
+#include "devicetree.h"
+#include "missing_efi.h"
+#include "util.h"
+
+#define FDT_V1_SIZE (7*4)
+
+#ifndef EFI_DT_FIXUP_PROTOCOL_GUID
+#define EFI_DT_FIXUP_PROTOCOL_GUID \
+        { 0xe617d64c, 0xfe08, 0x46da, {0xf4, 0xdc, 0xbb, 0xd5, 0x87, 0x0c, 0x73, 0x00} }
+#define EfiDtFixupProtocol ((EFI_GUID)EFI_DT_FIXUP_PROTOCOL_GUID)
+#endif
+
+#define EFI_DT_FIXUP_PROTOCOL_REVISION 0x00010000
+
+/* Add nodes and update properties */
+#define EFI_DT_APPLY_FIXUPS    0x00000001
+/*
+ * Reserve memory according to the /reserved-memory node
+ * and the memory reservation block
+ */
+#define EFI_DT_RESERVE_MEMORY  0x00000002
+
+typedef struct _EFI_DT_FIXUP_PROTOCOL EFI_DT_FIXUP_PROTOCOL;
+
+typedef EFI_STATUS (EFIAPI *EFI_DT_FIXUP) (
+        IN EFI_DT_FIXUP_PROTOCOL *This,
+        IN VOID                  *Fdt,
+        IN OUT UINTN             *BufferSize,
+        IN UINT32                Flags);
+
+struct _EFI_DT_FIXUP_PROTOCOL {
+        UINT64         Revision;
+        EFI_DT_FIXUP   Fixup;
+};
```

Comment:
shouldn't these definitions for the fixup protocol also be in missing_efi.h or so?

Reply:
(hmm, that's because this is a private u-boot api, i presume? i guess in that case it's ok)

Reply:
I'm not sure, but I think the idea is to get this protocol through the EFI process, but until that happens it's up to the EFI implementation I guess. I'm sure @xypron knows more.

Reply:
@poettering, @esmil Moving definitions into includes makes sense. The protocol specification is available in https://github.com/U-Boot-EFI/EFI_DT_FIXUP_PROTOCOL#readme. I have used this definition for the implementation in U-Boot and patches pending for GRUB.

Reply:
Cool. So there is a chance that it will eventually become part of the UEFI spec and implemented in gnu-efi. I'll move it to the misseng_efi.h file then, thanks.